### PR TITLE
Feature host groups 1

### PR DESCRIPTION
**Pull request by msekania**
Additional plugin that manages host groups. Created in close analogy with host.py and partly folders.py. In connection with rule module one can create a monitoring groups assigned to chosen subfolder.

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

-

## What is the new behavior?

- manage host groups in Checkmk (create/modify/delete)
- in connections with rule allows to monitor all hosts in the given subfolder

## Other information

For examle:

```yaml
- name: "Create a single host group."
  tribe29.checkmk.host_group:
    server_url: "http://localhost/"
    site: "my_site"
    automation_user: "automation"
    automation_secret: "$SECRET"
    host_group_name: "my_host_group"
    title: "My Host Group"
    state: "present"

- name: "Check_MK: create/register rules, which assign all servers in the given folder to the corresponding host group"
  tribe29.checkmk.rule:
    server_url: "http://localhost/"
    site: "my_site"
    automation_user: "automation"
    automation_secret: "$SECRET"
    ruleset: "host_groups"
    rule:
        folder: "/My/Host_Groups"
        conditions: {
            "host_labels": [],
            "host_tags": [],
            "service_labels": []
        }
        properties: {
            "disabled": false,
        }
        value_raw: "'my_host_group'"
    state: "present"
  run_once: 'true'
  delegate_to: localhost
```

### CODE REVIEW

**Comment by msekania** [comment]
simple module that manages host group, creates/deletes/modifies. created from host.py and partly folder.py

**Comment by msekania** [comment]
Hi @lgetwan,

No, problem, I'll take a look at tests/integration/targets/host. I will get back to you if I will have any questions.

Meanwhile, I have almost finished bulk_host_group module, which uses "bulk part" of api, but I will make a separate feature pull request for it. 

Is anyone working on contact groups? In our group we also plan to integrate contact group management in ansible, API is quite similar to host groups, and one can directly  adapt this two modules then.

Best regards,
Michael

**Comment by robin-tribe29** [comment]
Hi Michael, thanks for your contribution! :clap: 
Let me answer the question about contact groups: To our knowledge, no one is working on that feature, so would appreciate it, if you contributed something there too. :partying_face:

**Comment by msekania** [comment]
Hi @lgetwan,

test is included. I have also included parts for simple idem-potency tests for "present" and "absent".

Best regards,
Michael

File: tests/integration/targets/host_group/tasks/test.yml [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  run_once: true
+
+- name: "{{ outer_item.version }} - Modify part of host groups (also checks for idempotency!)."
+  host_group:
+    server_url: "{{ server_url }}"
+    site: "{{ outer_item.site }}"
+    automation_user: "{{ automation_user }}"
+    automation_secret: "{{ automation_secret }}"
+    host_name: "{{ item.name }}"
+    folder: "{{ item.title }}"
```
Comment (lgetwan) [inline tests/integration/targets/host_group/tasks/test.yml:37]: Same here

File: tests/integration/targets/host_group/tasks/test.yml [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  run_once: true
+
+- name: "{{ outer_item.version }} - Delete host groups."
+  host:
+    server_url: "{{ server_url }}"
+    site: "{{ outer_item.site }}"
+    automation_user: "{{ automation_user }}"
+    automation_secret: "{{ automation_secret }}"
+    host_name: "{{ item.name }}"
+    folder: "{{ item.title }}"
```
Comment (lgetwan) [inline tests/integration/targets/host_group/tasks/test.yml:65]: and here

File: tests/integration/targets/host_group/vars/main.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    site: "stable"
+  - version: "2.0.0p28"
+    site: "oldstable"
+download_url: "https://download.checkmk.com/checkmk/{{ item.version }}/check-mk-raw-{{ item.version }}_0.{{ ansible_distribution_release }}_amd64.deb"
+server_url: "http://127.0.0.1/"
+automation_user: "cmkadmin"
+automation_secret: "d7589df1-01db-4eda-9858-dbcff8d0c361"
+
+checkmk_host_groups_create:
+  - name:  "test1.tld"
```
Comment (lgetwan) [inline tests/integration/targets/host_group/vars/main.yml:13]: "." is not a valid character in a host group name.

File: tests/integration/targets/host_group/tasks/test.yml [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    automation_user: "{{ automation_user }}"
+    automation_secret: "{{ automation_secret }}"
+    force_foreign_changes: true
+    sites:
+      - "{{ outer_item.site }}"
+  delegate_to: localhost
+  run_once: true
+
+- name: "{{ outer_item.version }} - Delete host groups that were created at the beginning (also check for idempotency)."
+  host:
```
Comment (lgetwan) [inline tests/integration/targets/host_group/tasks/test.yml:84]: should be host_group

File: tests/integration/targets/host_group/tasks/test.yml [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    server_url: "{{ server_url }}"
+    site: "{{ outer_item.site }}"
+    automation_user: "{{ automation_user }}"
+    automation_secret: "{{ automation_secret }}"
+    host_name: "{{ item.name }}"
+    folder: "{{ item.title }}"
```
Comment (lgetwan) [inline tests/integration/targets/host_group/tasks/test.yml:90]: host_group_name & title

**Review by lgetwan — CHANGES_REQUESTED** [review]
The module itself looks fine. Please try out the integration test using:
`ansible-test integration --docker default host_group`

File: tests/integration/targets/host_group/tasks/test.yml (hunk 5 above)
Comment (msekania) [reply tests/integration/targets/host_group/tasks/test.yml:90]: I'll fix the file

File: tests/integration/targets/host_group/tasks/test.yml (hunk 1 above)
Comment (msekania) [reply tests/integration/targets/host_group/tasks/test.yml]: should be host_group_name & title

File: tests/integration/targets/host_group/tasks/test.yml (hunk 2 above)
Comment (msekania) [reply tests/integration/targets/host_group/tasks/test.yml]: should be host_group_name & title

**Comment by msekania** [comment]
thanks! I had to be more careful :) 
I have made the changes.

File: tests/integration/targets/host_group/vars/main.yml (hunk 3 above)
Comment (msekania) [reply tests/integration/targets/host_group/vars/main.yml]: Yes

**Review by msekania — COMMENTED** [review]
Thanks!
Corresponding changes have been made.

File: tests/integration/targets/host_group/tasks/test.yml [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+  host_group:
+    server_url: "{{ server_url }}"
+    site: "{{ outer_item.site }}"
+    automation_user: "{{ automation_user }}"
+    automation_secret: "{{ automation_secret }}"
+    host_group_name: "{{ item.name }}"
+    title: "{{ item.title }}"
+    attributes:
+      site: "{{ outer_item.site }}"
+      ipaddress: 127.0.0.1
```
Comment (lgetwan) [inline tests/integration/targets/host_group/tasks/test.yml:12]: These lines are only valid with the host or folder module.

File: tests/integration/targets/host_group/tasks/test.yml [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
```
Comment (lgetwan) [inline tests/integration/targets/host_group/tasks/test.yml:40]: Same here.

**Review by lgetwan — CHANGES_REQUESTED** [review]
Hi Michael,
there are still two remains from the host module in your integration test playbook. Once they are removed, it runs like a charm.
Best regards
Lars

File: tests/integration/targets/host_group/tasks/test.yml (hunk 7 above)
Comment (msekania) [reply tests/integration/targets/host_group/tasks/test.yml]: removed

File: tests/integration/targets/host_group/tasks/test.yml (hunk 6 above)
Comment (msekania) [reply tests/integration/targets/host_group/tasks/test.yml]: removed

**Comment by robin-tribe29** [comment]
Hi Micheal,

we are targeting #141 for the upcoming release, so I will merge this PR for the 0.12.0 release, pending a quick review of my own.
Thanks a lot for your contribution and your patience, we do appreciate it!

**Comment by robin-tribe29** [comment]
I just realized, that you put the module file in the root of the project, rather than `plugins/modules` @msekania, please fix that.

**Comment by msekania** [comment]
@robin-tribe29 bulk host groups is now integrated in host_group module. tests are also extended.

**Review by lgetwan — APPROVED** [review]
Hi Michael,

looks very good from my point of view!
Thanks a lot for your work! :-)

Best regards
Lars

**Comment by msekania** [comment]
Hi Lars,

> looks very good from my point of view!
Thanks!

There is one minor issue which I have noticed during contact_group.py generation.
In host_group.py module, in the examples part, I have overlooked "hostgroups".
In lines 70, 86, and 110, "hostgroups" should be updated to "host_groups"

Thanks, and best regards
Michael

**Comment by robin-tribe29** [comment]
Fixed in 4d09a6cf38ada63ff49d99f29bbfa029b5fe1dd5.

**Pull request merged**